### ISSUE TITLE
The ufuncify numpy and cython backends may have too much overhead

### ISSUE THREAD

**Issue by moorepants** [sympy/sympy#8997]
Check out this gist:

http://nbviewer.ipython.org/gist/moorepants/ae16ef9d87eca0af72a9

It compares numba, numpy, and ufuncify. Our numpy and cython backends must have too much overhead as they should be comparable to the f2py backend and numba. They should definitely be faster than the pure numpy, but they aren't.

**Comment by jcrist** [comment]
This is just the pow evaluation issue again. C doesn't have fast evaluation for integer powers (always uses full `pow` machinery, instead of converting to x*x), like fortran does. If we used `-ffast-math`, C should do those optimizations.

**Comment by moorepants** [comment]
Ah! I've had trouble compiling with `-ffast-math` in my opty package so I disabled it.

I'll try running the benchmark with `-ffast-math`. We should change the C code printer to do this. Is there any reason not to?

**Comment by moorepants** [comment]
It doesn't seem like the `CythonCodeWrapper` class has easy support for adding extra compile flags.

**Comment by jcrist** [comment]
Because floating point math is non-associative, this can lead to increased errors in calculations, that the user may not want. Last time we discussed this, a few options were suggested:

1. Turn on `-ffast-math`
2. Print integer powers less than some number as optimally parenthesized multiplications
3. Link in our own `powi` implementation for integer powers.

I'm not sure what's best here. I dislike number 2, as it's an optimization that I don't think should be done at this level, and it prevents user overriding of integer pow. Option 3 would be a sharp change, but is the best option if we don't want to turn on an optimization that the user may not want.

**Comment by jcrist** [comment]
> It doesn't seem like the CythonCodeWrapper class has easy support for adding extra compile flags.

Yeah, we never did get around to that. For a quick test, you could generate the code with a specified file, manually edit it, then recompile.

**Comment by moorepants** [comment]
Relevant issues:
- https://github.com/sympy/sympy/pull/7517
- https://github.com/sympy/sympy/pull/7519

**Comment by moorepants** [comment]
@jcrist Just added this: https://github.com/sympy/sympy/pull/9001

But I don't get any speed ups:

http://nbviewer.ipython.org/gist/moorepants/a1ba433f9437602a0d2f

**Comment by moorepants** [comment]
Comments:
> 1. Turn on `-ffast-math`
This should be an option regardless if we change the printing behavior.
> 2. Print integer powers less than some number as optimally parenthesized multiplications
I don't see anything fundamentally bad about this. You can also say that we shouldn't do CSE at this level and leave that to the compiler, but I'm not sure the compilers always do the best thing. I'm pretty sure that pre-CSEing the code before and optimized compile results in faster code. I'll implement this in PyDy to try to prove it.
3. Link in our own `powi` implementation for integer powers.
Once again, we should probably support this functionality too because it could be useful for other functions.

**Comment by jcrist** [comment]
Hmmm, I'm not sure then. When I was working on this exact same issue this summer, I found that I could make C and Fortran almost equivalent by replacing `pow` with `x*x...`. The `-ffast-math` flag *should* perform that optimization for us here, but it may not be. I don't have time to dig into this today, but I'll try to look at it this week.

**Comment by moorepants** [comment]
I like this comment: http://stackoverflow.com/a/6442503/467314

It says "if you want x*x*x*x, then write that, don't write pow(x)". So I think we are fine writing that with our printers, but of course we are messing with the floating point accuracy. The question we have to ask is if a sympy expression orders muls in a certain way or has a pow, then what should we be printing on the code printers? I'm guessing this is problem dependent. I'm not sure how we translate from theorectical expressions to floating point calcs. For example, something that comes out of `KanesMethod` or `LangrangesMethod` will have these multiplications chosen by the underlying SymPy algorithms. So we depend on SymPy to "choose" the right form, or we retain the form that the user inputs. If a user makes a sympy expression: `x * x * x` then SymPy is already screwing it up for down the road floating point:

```
In [1]: import sympy as sm

In [2]: x = sm.Symbol('x')

In [3]: x * x * x
Out[3]: x**3

In [4]: type(x * x * x)
Out[4]: sympy.core.power.Pow
```

So this is a deeper issue. But I think we have the right to manipulate these expressions with our printers to make faster code. The bad news is we have no idea what the expression creator intended. So, by default, SymPy is not a suitable platform for retaining floating point operation order.

@certik I'd be curious if you have any comments on this.

**Comment by moorepants** [comment]
Maybe the best is to have a "low_accuracy" flag for the C printer. Then we can make any kind of substitutions we want and disregard the associativity limit on floating point operations. The user will then know that they may get fast code, but it may be less accurate.

**Comment by certik** [comment]
Mathematically, `x*x*x = x^3`, so there is no issue for SymPy doing this simplification. You can use the `hold` attribute to represent `x*x*x`.

I would just use `-ffast-math` (the other options that I use are `-O3 -march=native -ffast-math -funroll-loops`). It seems to me the precision is dropped only about 4x or so in that stackoverflow post. I think that's negligible, but of course, ultimately the user will have to make that call.

**Comment by moorepants** [comment]
> Mathematically, `x*x*x = x^3`, so there is no issue for SymPy doing this simplification. You can use the `hold` attribute to represent `x*x*x`.

It isn't an issue mathematically but if you are specifically targeting floating point operations with the SymPy expressions, it can be a issue. I'm not sure how we could apply a `hold` throughout all of the computations that happen in the mechanics package, or if we'd want to.

Thanks for the compiler flag tips. It isn't clear that `fast-math` actually "unrolls" the low integer `pow()` calls. I didn't get a speed up above in the example I posted. There seem to be things that we can do at the SymPy printing level that the compiler optimizers may not do, or maybe even SymPy can do them faster/better...not sure though.

**Comment by certik** [comment]
The printing should definitely have an option to use the most optimized expression so that the gcc can produce fast code. If manually unrolling powers is the only way to get fast code, then we should support it. I am just surprised that this is the case. What compiler do you use?

**Comment by moorepants** [comment]
I use gcc most of the time. I haven't done extensive testing on the compiler flags, so there probably isn't any reason to discuss further until we do. I'll try to run some more tests this week when I have some free time.

**Comment by certik** [comment]
Cool. Make sure you enable `-march=native`, that's essential for good performance.

**Comment by Wrzlprmft** [comment]
For your interest, the optimisation of small integer powers is handled by the following flags:

- GCC: `-funsafe-math-optimizations` which is activated by `ffast-math`, which in turn is activated by `-Ofast`
- Clang 3.8: `ffast-math` which is activated by `-Ofast`.
- Clang 3.4: I could not find any such flag.

If this optimisation is activated, `pow(x,3)` is computed at least ten times as fast.

I also wish to remark that an option for the C code printer to use optimised expressions would come quite in handy for [this project of mine](http://github.com/neurophysik/jitcode).

**Comment by bjodah** [comment]
Just a small remark: `pow(x, 2)` generates the same assembly as `x*x` using gcc 5.4.0 when `-O1` is given (not for `-O0`). The unsafe simplification for `pow(x, 3)` still requires an active choice by setting the `-funsafe-math-optimizations` flag (together with `-O1`).

Bash script to check this (if someone is interested):

```bash
#!/bin/bash
function get_assembly() {
    gcc -O1 -S -x c -o - - <<EOF
#include <math.h>
double f(double x) { return $1; }
EOF
}
result_pow=$(get_assembly "pow(x, 2)")
result_expanded=$(get_assembly "x*x")
diff <(echo $result_pow) <(echo $result_expanded)
```

**Comment by asmeurer** [comment]
But `-O0` is quite common for generated code. `-O1` ranges from taking ~hours to compile to just never finishing (or crashing the compiler) for large code generated expressions. Even if you can split out the proper flag from the compiler, it would be much simpler to just output things better in the first place.
